protocol/communities/manager.go

```diff
@@ -2320,39 +2320,41 @@ func (m *Manager) accountsSatisfyPermissionsToJoin(community *Community, account
 }
 
 func (m *Manager) accountsSatisfyPermissionsToJoinChannels(community *Community, accounts []*protobuf.RevealedAccount) (map[string]*protobuf.CommunityChat, map[string]*protobuf.CommunityChat, error) {
-	accountsAndChainIDs := revealedAccountsToAccountsAndChainIDsCombination(accounts)
+	viewChats := make(map[string]*protobuf.CommunityChat)
+	viewAndPostChats := make(map[string]*protobuf.CommunityChat)
 
-	viewChats, err := m.accountsSatisfyPermissionTypeToJoinChannels(community, accountsAndChainIDs, protobuf.CommunityTokenPermission_CAN_VIEW_CHANNEL)
-	if err != nil {
-		return nil, nil, err
-	}
+	accountsAndChainIDs := revealedAccountsToAccountsAndChainIDsCombination(accounts)
 
-	postChats, err := m.accountsSatisfyPermissionTypeToJoinChannels(community, accountsAndChainIDs, protobuf.CommunityTokenPermission_CAN_VIEW_AND_POST_CHANNEL)
-	if err != nil {
-		return nil, nil, err
-	}
+	for channelID, channel := range community.config.CommunityDescription.Chats {
+		channelViewOnlyPermissions := community.ChannelTokenPermissionsByType(community.IDString()+channelID, protobuf.CommunityTokenPermission_CAN_VIEW_CHANNEL)
+		channelViewAndPostPermissions := community.ChannelTokenPermissionsByType(community.IDString()+channelID, protobuf.CommunityTokenPermission_CAN_VIEW_AND_POST_CHANNEL)
+		channelPermissions := append(channelViewOnlyPermissions, channelViewAndPostPermissions...)
 
-	return viewChats, postChats, nil
-}
+		if len(channelPermissions) == 0 {
+			viewAndPostChats[channelID] = channel
+			continue
+		}
 
-func (m *Manager) accountsSatisfyPermissionTypeToJoinChannels(community *Community, accounts []*AccountChainIDsCombination, permissionType protobuf.CommunityTokenPermission_Type) (map[string]*protobuf.CommunityChat, error) {
-	result := make(map[string]*protobuf.CommunityChat)
+		permissionResponse, err := m.PermissionChecker.CheckPermissions(channelPermissions, accountsAndChainIDs, true)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	for channelID, channel := range community.config.CommunityDescription.Chats {
-		permissions := community.ChannelTokenPermissionsByType(community.IDString()+channelID, permissionType)
-		if len(permissions) > 0 {
-			permissionResponse, err := m.PermissionChecker.CheckPermissions(permissions, accounts, true)
-			if err != nil {
-				return nil, err
+		if permissionResponse.Satisfied {
+			highestRole := calculateRolesAndHighestRole(permissionResponse.Permissions).HighestRole
+			if highestRole == nil {
+				return nil, nil, errors.New("failed to calculate highest role")
 			}
-			if !permissionResponse.Satisfied {
-				continue
+			switch highestRole.Role {
+			case protobuf.CommunityTokenPermission_CAN_VIEW_CHANNEL:
+				viewChats[channelID] = channel
+			case protobuf.CommunityTokenPermission_CAN_VIEW_AND_POST_CHANNEL:
```

Comment:
I don't understand this code very well, so is there a case where the highestRole is Owner? If so, should we also add it as `viewAndPostChats`?

Reply:
Not in this case, in the scope of this function (`accountsSatisfyPermissionsToJoinChannels`) only channel permissions are evaluated.